sync/src/orphan_block_pool.rs

```diff
@@ -1,152 +1,326 @@
-use ckb_logger::debug;
+use std::collections::{HashMap, VecDeque};
+use std::ops::Deref;
+
+use ckb_error::{Error, InternalErrorKind};
+use ckb_types::prelude::ShouldBeOk;
 use ckb_types::{core, packed};
-use ckb_util::{parking_lot::RwLock, shrink_to_fit};
-use std::collections::{HashMap, HashSet, VecDeque};
+use ckb_util::parking_lot::RwLock;
+use ckb_util::shrink_to_fit;
 
+/// alias name
 pub type ParentHash = packed::Byte32;
+/// alias name
+pub type Hash = packed::Byte32;
+
 const SHRINK_THRESHOLD: usize = 100;
+const MAX_SUB_NODES_UPDATE_IN_ROOTOF: usize = 16;
+
+/// RwLock of wrapper for OrphanBlockPool
+pub struct OrphanBlockPool {
+    inner: RwLock<InnerOrphanBlockPool>,
+}
+
+impl OrphanBlockPool {
+    ///  init orphan block pool, reserve maps with specific capacity
+    pub fn with_capacity(capacity: usize) -> Self {
+        OrphanBlockPool {
+            inner: RwLock::new(InnerOrphanBlockPool::with_capacity(capacity)),
+        }
+    }
+    /// insert the orphan block into pool
+    ///
+    /// Return Ok() normally
+    ///
+    /// Return Error when inner data not sync
+    pub fn insert(&self, block: core::BlockView) -> Result<(), Error> {
+        self.inner.write().insert(block)
+    }
+    /// return vector of sub-node blocks beneath to the root hash
+    ///
+    /// return None if none beneath the root
+    pub fn remove_blocks_by_parent(&self, root: &Hash) -> Vec<core::BlockView> {
+        self.inner
+            .write()
+            .remove_blocks_by_parent(root)
+            .unwrap_or_default()
+    }
+    /// return block stored in pool
+    ///
+    /// return None if not found in pool
+    pub fn get_block(&self, hash: &Hash) -> Option<core::BlockView> {
```

Comment:
Can we avoid this function? It has a worse performance in the new data strcuture.

Reply:
`get_block` function is already existed, and used in two places: one for test code(ok, omit it); the second is "sync/src/relay/mod.rs", function `reconstruct_block()`

get_block itself is simple(but worse performance ) and probably not improved by refactor,  so there is only way: NOT used it. 
i will check `reconstruct_block()` if possible

Reply:
in reconstruct_block(), the context of use `get_block()` :  
when the code goes to get block of 'uncle hash', it checks state.block_status_map to check if it received....
ok, pool already received it before, so let fetch it:
self.shared.state().get_orphan_block(&uncle_hash)  --> it will call orphan pool's `get_block()`

i **suspect** much the uncle block is  trees.values().front()  (1st node of every value of trees)
if it's true, we could change it into simple and fast implementation.
but i don't have enough knowledge to answer this question or confirm it, i need help.

Reply:
in new design, there are 2 tables:  
- root_of: keeps relationship(child -->father or ancestor)
- trees: keeps ancestor and all_connected_children block data

as above design, there is no shortcut, if we want to fetch any block.

Reply:
> i suspect much the uncle block is trees.values().front() (1st node of every value of trees)

This assumption is wrong; any block in the current epoch can be accepted as an uncle block

Reply:
so we have to use `get_block`...i have no clue to improve it, any suggestions?